src/core/experimental/refcount.d

```diff
@@ -0,0 +1,390 @@
+    {
+        mixin(opAssignImpl);
+    }
+
+    private enum opAssignImpl =
+    q{
+        if (rhs.isInitialized())
+        {
+            if (rc == rhs.rc)
+            {
+                return this;
+            }
+            rhs.addRef();
+        }
+        if (isInitialized())
+        {
+            delRef();
+        }
+        () @trusted { rc = rhs.rc; }();
+        return this;
+    };
+
+    /*
+    Increase the reference count. This asserts that `__RefCount` is initialized.
+
+    Returns:
+         This returns a `void*` so the compiler won't optimize away the call
+         to this `const pure` function.
+    */
+    @nogc nothrow pure @safe scope
+    private void* addRef(this Q)() const
+    {
+        assert(isInitialized(), "__RefCount is uninitialized");
+        cast(void) rcOp!(Q, "+=")(1);
+        return null;
+    }
+
+    /*
+    Decrease the reference count. If this was the last reference, `free` the
+    support. This asserts that `__RefCount` is initialized.
+
+    Returns:
+         This returns a `void*` so the compiler won't optimize away the call
+         to this `const pure` function.
+    */
+    @nogc nothrow pure @trusted scope
+    private void* delRef(this Q)() const
+    {
+        assert(isInitialized(), "__RefCount is uninitialized");
+        /*
+         * This is an optimization. Most likely, most of the time, the refcount
+         * is `1`, so we don't want to make more ops to update that value only
+         * to free it afterwards. This is why we first check if the value is `1`.
+         * The `or` part decrements and checks for `0` in the case that we had
+         * two threads that both questioned at the same time if the value was `1`;
+         * they both will decrement, but only one will have the final value `0`.
+         */
+        if (rcOp!(Q, "==")(1) || (rcOp!(Q, "-=")(1) == 0))
+        {
+            return deallocate();
+        }
+        return null;
+    }
+
+    /*
+    `free` the support.
+
+    Returns:
+         This returns a $(B void*) so the compiler won't optimize away the call
+         to this `const pure` function.
+    */
+    @nogc nothrow pure @system scope
+    private void* deallocate(this Q)() const
+    {
+        import core.memory : pureFree;
+        pureFree(cast(void*) rc);
+        return null;
+    }
+
+    /**
+    Destruct the `__RefCount`. If it's initialized, decrement the refcount.
+    */
+    @nogc nothrow pure @trusted scope
+    ~this()
+    {
+        if (isInitialized())
+        {
+            delRef();
+        }
+    }
+
+    /**
+    Return a boolean value denoting if this is the only reference to this object.
+
+    Returns:
+         `true` if this reference count is unique; `false` if this `__RefCount`
+         object is uninitialized or there are multiple references to it.
+
+    Complexity:
+         $(BIGOH 1).
+    */
+    pure nothrow @safe @nogc scope
+    bool isUnique(this Q)() const
```

Comment:
This function must be removed. I can't imagine anything you could do with this that's not a data race.

Reply:
Removing the call to `isInitialized` will remove the data race at the cost of generating a `segfault` for an invalid (uninitialized) `__RefCount`ed instance.

This function provides a simple interface towards the user to help him decide if he needs to deallocate the internal state of his data structure.
An "alternative" to this would be to expose the internal `delRef` to the user, which could break / invalidate the automatic reference counting. The user should never be able to explicitly modify the RC, but only "view" the value.

I propose that we bite the bullet and take the segfault.

Reply:
If the RC is atomic, then none of those things you say can exist. Everything you just said is a data race.
If RC is thread-local (as it should be in D), then we have options.